tests/framework/prototype.py

```diff
@@ -38,41 +61,82 @@ def search_path(yaml_object, steps, default_value = None):
             return default_value
     return curr
 
-# Functions for cluster preparation.
-def delete_kind_cluster() -> None:
-    """Delete a KinD cluster"""
-    shell_subprocess_run("kind delete cluster")
-
-def create_kind_cluster():
-    """Create a KinD cluster"""
-    shell_subprocess_run("kind create cluster --wait 900s")
-
-def install_crd():
-    """Install Custom Resource Definition (CRD)"""
-    shell_subprocess_run("kubectl create -k ../../manifests/cluster-scope-resources")
-
-def download_images(docker_images):
-    """Download Docker images from DockerHub"""
-    docker_client = docker.from_env()
-    for image in docker_images:
-        # Only pull the image from DockerHub when the image does not
-        # exist in the local docker registry.
-        if shell_subprocess_run(f'docker image inspect {image} > /dev/null', check = False) != 0:
-            docker_client.images.pull(image)
-    docker_client.close()
-
-def kind_load_images(docker_images):
-    """Load downloaded images into KinD cluster"""
-    for image in docker_images:
-        shell_subprocess_run(f'kind load docker-image {image}')
-
-def install_kuberay_operator():
-    """Install kuberay operator with image kuberay/operator:nightly"""
-    shell_subprocess_run('kubectl apply -k ../../manifests/base')
-
-def check_cluster_exist():
-    """Check whether KinD cluster exists or not"""
-    return shell_subprocess_run("kubectl cluster-info --context kind-kind", check = False) == 0
+class KubernetesClusterManager:
+    """
+    KubernetesClusterManager controlls the lifecycle of KinD cluster and Kubernetes API client.
+    """
+    def __init__(self) -> None:
+        self.k8s_client_dict = {}
+
+    def delete_kind_cluster(self) -> None:
+        """Delete a KinD cluster"""
+        shell_subprocess_run("kind delete cluster")
+        for _, k8s_client in self.k8s_client_dict.items():
+            k8s_client.api_client.rest_client.pool_manager.clear()
+            k8s_client.api_client.close()
+        self.k8s_client_dict = {}
+
+    def create_kind_cluster(self, kind_config = None) -> None:
+        """Create a KinD cluster"""
+        # To use NodePort service, `kind_config` needs to set `extraPortMappings` properly.
+        kind_config = CONST.DEFAULT_KIND_CONFIG if not kind_config else kind_config
+        shell_subprocess_run(f"kind create cluster --wait 900s --config {kind_config}")
+        config.load_kube_config()
+        self.k8s_client_dict.update({
+            CONST.K8S_V1_CLIENT_KEY: client.CoreV1Api(),
+            CONST.K8S_CR_CLIENT_KEY: client.CustomObjectsApi()
+        })
+
+    def check_cluster_exist(self) -> bool:
+        """Check whether KinD cluster exists or not"""
+        return shell_subprocess_run("kubectl cluster-info --context kind-kind", check = False) == 0
+K8S_CLUSTER_MANAGER = KubernetesClusterManager()
```

Comment:
There should be a space (or two) after the class. 
Would you mind running a linter?

Reply:
My pylint is always running. Would you mind explaining "There should be a space (or two) after the class." more? Do you mean add newlines between L92 and L93? Thank you!

Reply:
Yeah, that's what I meant. Maybe our linters are different.

Instantiating the class in the line after its defined makes enough sense to me!  Don't mind me.